### ISSUE TITLE
Rewriting Users

### ISSUE THREAD

**Issue by QuiCM** [Pryaxis/TShock#1513]
Users have problems that need fixing, see the attached Project for more details.

This issue is for discussion on how we should go about fixing the mentioned issues

**Comment by hakusaro** [comment]
I would add that we have a lot of bad language. [Player] and [User] really need to be disambiguated so that it's clear that hey, this is something [active on the server playing] and this is the [data store about their account with us].

**Comment by QuiCM** [comment]
Perhaps a rename from User to Account

**Comment by Ryozuki** [comment]
"Relational mappings & sane data typing" (https://github.com/Pryaxis/TShock/projects)
Does this mean moving all the methods that query/insert to each respective class? 
A.k.a getting rid of Database managers such as CharacterManager?

E.g
```csharp
// CharacterManager.cs
public PlayerData GetPlayerData(TSPlayer player, int acctid)
// to
class TSPlayer 
{
    // ...
    public void GetData(); // uses the already stored id to retrieve data from db
}
```

**Comment by ivanbiljan** [comment]
@Ryozuki no. Relational mappings address the issue of the current database design (IPs are stored in a single column as a serialized array, we assume that the 'Group' column matches the Group table...). You should read up a bit on database normalization.
A more acceptable design should look something like this (in my humble opinion):

The 'Users' table: should contain basic information about an account. This includes:
   - **AccountID**: the unique account identifier
   - **Username**: the account's name (should probably be PKd)
   - **UUID**: the account's UUID 
   - **Register Date**: the account's registration date
   - **Last Accessed**: the date that the account was last accessed (logged in to)

The 'UserHasGroup' table: should link the two tables ('Users' and 'Groups') 
   - **User**: foreign key. Holds a reference to the corresponding account in the 'Users' table
   - **Group**: foreign key. Holds a reference to the corresponding group in the 'Groups' table

Similarly we should have a 'UserHasIP' table or such. This is a simple example I came up with to clarify the meaning of "relational mappings", you're free to use your own ideas.

**Comment by QuiCM** [comment]
Blocking this on fragments too
